### PR TITLE
WIP python-sgd-regression

### PR DESCRIPTION
**Pull request by Marigold**
First attempt at distributed `python-sgd-regression`. It's incomplete and not functional at the moment, I just need it for feedback from @ludovicc.

The code contains a very rough structure (which could be very likely wrong, in that case please correct me!) for distributed [SGDRegressor](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.SGDRegressor.html). Questions are labeled with `# QUESTION` comment, it would be great to have your advice there @ludovicc. Btw some of them might be pretty stupid, but I'd rather ask than assume at this point.

Some of the `# TODO`s for this PR need to be done in [io_helper](https://github.com/HBPMedical/python-base-docker-images/blob/master/python-mip/io_helper/io_helper/io_helper.py) module.

By the way, any ideas how to DRY python code in `algorithm-repository`? `python-linear-regression` will share a lot of python code with `python-sgd-regression` (and `python-sgd-classifier` later), so it would make sense to have all this in once place. I was thinking about

1. putting this code to `python-mip` docker image to `io_helpers` library or a new one
2. creating a new util-library directly in `algorithm-repository` and adding it to Docker as a requirement

What do you think makes more sense? 2. has the advantage of being close to the code (and having to create only one PR when making changes), but there might be some issues I'm not seeing right now

### CODE REVIEW

File: python-sgd-regression/db_models.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# TODO: this module with ORM for database should go to io_helpers
```
Comment (ludovicc) [inline python-sgd-regression/db_models.py]: yes indeed

File: python-sgd-regression/db_models.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+# QUESTION: does it make sense to use heavy ORM like SQLALchemy or should I simply use raw SQL through SqlAlchemy? (or
+# maybe different ORM?)
+
```
Comment (ludovicc) [inline python-sgd-regression/db_models.py]: Let's use SQLAlchemy for JobResult. Otherwise I'm happy with SqlAlchemy + raw SQL

File: python-sgd-regression/sgd_regression.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
+#!/usr/bin/env python3
+
+# QUESTION: when creating files in python-sgd-regression I copied folder python-linear-regression and just renamed
+# `linear-regression` to `sgd-regression`. Is there another way? e.g. cookiecutter template or something?
+
```
Comment (ludovicc) [inline python-sgd-regression/sgd_regression.py]: A template would be awesome, but there are many variants to manage (Java, Python, Python + Scikit...) and we are still changing things. I'm starting to use Atomist.com to organise the projects, but creating templates is a hassle for far.

File: python-sgd-regression/sgd_regression.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+# QUESTION: I'd like to create some python unit tests in py.test (to speed up development), any suggestions where should
+# I put them?
+
```
Comment (ludovicc) [inline python-sgd-regression/sgd_regression.py]: I'm starting to add unit tests as well. Check this: https://github.com/LREN-CHUV/algorithm-repository/tree/feature/test-histograms/python-histograms/tests

File: python-sgd-regression/sgd_regression.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+def deserialize_sklearn_estimator(js):
+    """Init model from JSON."""
+    raise NotImplementedError()
+
+
+def sklearn_to_pfa(estimator):
+    """Convert scikit-learn to PFA."""
+    # QUESTION: What's the best way for generating PFA code from scikit models in your opinion? Should I use PrettyPFA
```
Comment (ludovicc) [inline python-sgd-regression/sgd_regression.py]: Can you try using PrettyPFA? It should work well from Python

File: python-sgd-regression/sgd_regression.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+    job_id = io_helper._get_job_id()
+    partial_job_id = None
+
+    # Start reading stream of inputs
+    # QUESTION: the tables `cde_*` are just for a proof of concept? How is the data going to be distributed in
+    # production? Probably not as tables in a single database (in that case we wouldn't need all the serialization
+    # of models and storing intermediate results in `job_result` table). In separate databases? Or is this just a
+    # preparation for federated learning where each docker container (running on hospital's servers)
+    # will save its intermediate results? Thanks for explanation!
+
```
Comment (ludovicc) [inline python-sgd-regression/sgd_regression.py]: The tables `cde_*` are just for a proof of concept.

On each hospital, there is a database containing currently one table (mip_cde_features), but there will be variants. Anyway, data is located on different databases that cannot talk directly to each other, so we need to perform the learning on each of them, and gather the results using web services. This is done by Woken.

File: python-sgd-regression/sgd_regression.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+    # QUESTION: related to the question above - if we indeed have all the tables in a single DB, why do we have to
+    # store intermediate results to `job_result` table? Can't we just keep everything in memory?
+
```
Comment (ludovicc) [inline python-sgd-regression/sgd_regression.py]: Intermediate results are going to be copied from one hospital to another, and Woken will store the results gathered from a previous hospital into the job_results table of the local hospital before calling the algorithm.

File: wip/README.md [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+# How it all works:
+1. User specifices algorithm, variables & covariables in MIP frontend (creates JSON request in the background)
+2. POST request to `/mining/job` with body
+```
+{
+    "user": {"code": "user1"},
+    "variables": [{"code": "var1"}],
+    "covariables": [{"code": "var2"},{"code": "var3"}],
+    "grouping": [{"code": "var4"}],
+    "filters": [],
```
Comment (ludovicc) [inline wip/README.md]: targetTable: "cde_features_A",

File: wip/README.md [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+    "algorithm": "python-sgd-regression",
+    "datasets": [{"code": "dataset1"},{"code": "dataset2"}]
+}
+```
+3. `woken` converts request body to ENV variables and runs `docker-compose run python-sgd-regression compute`
+4. script in a container loads data from sample / production database tables `cde_features_A, cde_features_B and cde_features_C` and writes intermediate result to DB on each partial update (everything in that single script)
```
Comment (ludovicc) [inline wip/README.md]: The algorithm in a container should be called 3 times: once for cde_features_A, then cde_features_B and finally cde_features_C

File: python-sgd-regression/sgd_regression.py [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+    # TODO: we'd want to perform partial_fit multiple times on each database
+    # TODO: choose table randomly on each iteration (wrap it in a generator)
```
Comment (ludovicc) [inline python-sgd-regression/sgd_regression.py]: No, the table should be selected
- in Woken, using parameter targetTable in the json
- in the container, using PARAM_query environment variable with
  SELECT a,b,c from TARGET_TABLE

There should be 3 calls to the algorithm, one for each target table.

Comment (ludovicc) [reply python-sgd-regression/sgd_regression.py]: At the end of each call, the algorithm should store its partial model into job_result table, and this partial model should be loaded by the next algorithm

File: python-sgd-regression/tests/docker-compose.yml [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+      FLYWAY_DBMS: postgresql
+      FLYWAY_HOST: db
+      FLYWAY_PORT: 5432
+      FLYWAY_DATABASE_NAME: woken
+      FLYWAY_USER: postgres
+      FLYWAY_PASSWORD: test
+    depends_on:
+      - db
+
+  sgd-regression:
```
Comment (ludovicc) [inline python-sgd-regression/tests/docker-compose.yml]: sg-regression-a:
  environment:
   PARAM_query: select c1,c2,c3 from cde_features_a

sg-regression-b:
  environment:
   PARAM_query: select c1,c2,c3 from cde_features_b

sg-regression-c:
  environment:
   PARAM_query: select c1,c2,c3 from cde_features_c

File: python-sgd-regression/tests/test.sh [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+$DOCKER_COMPOSE run create_dbs
+
+echo
+echo "Initialise the databases..."
+$DOCKER_COMPOSE run sample_data_db_setup
+$DOCKER_COMPOSE run woken_db_setup
+
+echo
+echo "Run the sgd-regression algorithm..."
+$DOCKER_COMPOSE run sgd-regression compute
```
Comment (ludovicc) [inline python-sgd-regression/tests/test.sh]: $DOCKER_COMPOSE run sgd-regression-a compute
$DOCKER_COMPOSE run sgd-regression-b compute
$DOCKER_COMPOSE run sgd-regression-c compute

File: python-sgd-regression/sgd_regression.py (hunk 3 above)
Comment (Marigold) [reply python-sgd-regression/sgd_regression.py]: Agree. Didn't know about atomist, looks neat.

File: python-sgd-regression/sgd_regression.py (hunk 4 above)
Comment (Marigold) [reply python-sgd-regression/sgd_regression.py]: Oh nice. I'll reuse it in my code. Btw have you tried [py.test](https://docs.pytest.org/en/latest/) instead of python unittest library? It has made testing & development so much easier and enjoyable for me. There's just so much boilerplate code in standard python unittests.

File: python-sgd-regression/sgd_regression.py (hunk 10 above)
Comment (Marigold) [reply python-sgd-regression/sgd_regression.py]: > in the container, using PARAM_query environment variable with SELECT a,b,c from TARGET_TABLE

Thanks, it makes sense now.


> At the end of each call, the algorithm should store its partial model into job_result table, and this partial model should be loaded by the next algorithm

Is ENV `JOB_ID` going to change between runs? I can imagine two designs

### 1. `JOB_ID` will be the same
Each run will have to save both partial model (i.e. serialized sklearn model) and its PFA version (because the run doesn't know whether it is the last or not). Or somehow add meta parameters from regression to PFA that would allow its reconstruction in next run (like model's coefficients, learning rate, etc.).
Each run will UPDATE job_result table.

### 2. `JOB_ID` will change between runs
In that case we would save serialized runs in each but the last run which would save PFA.
Each run will INSERT into job_result table.

Could you please point me to woken code that is responsible for the iterative learning? I found something, but I'd rather ask. Thanks!

File: python-sgd-regression/sgd_regression.py (hunk 4 above)
Comment (ludovicc) [reply python-sgd-regression/sgd_regression.py]: We are using Nose (https://nose.readthedocs.io/). It also claims to make testing easier...

File: python-sgd-regression/sgd_regression.py (hunk 10 above)
Comment (ludovicc) [reply python-sgd-regression/sgd_regression.py]: There's no implementation yet for iterative learning. But all it needs is saving job results and moving them across nodes. There should be one master job_id, then a variant for each node, for example job_id = 1234 on master, then 1234-node1, 1234-node2

Comment (Marigold) [reply python-sgd-regression/sgd_regression.py]: The thing is that there's only one `JOB_ID` env variable. I would be expecting two job ids - one for the previous model (e.g. `1234-node1`) and one where should I save the new job result (e.g. `1234-node2`). Also an indication whether this is the last iteration (so that I can generate PFA) would be useful.

Anyway, the fact that it's not yet ready in woken doesn't really block me. I'll be just returning both PFA code and serialized scikit model for the time being. It should be easy to modify it when woken is ready for iterative learning.

File: python-sgd-regression/sgd_regression.py (hunk 4 above)
Comment (Marigold) [reply python-sgd-regression/sgd_regression.py]: Can I try pytest first if you don't mind and in case you don't like it, switch to nose?

File: python-sgd-regression/sgd_regression.py (hunk 5 above)
Comment (Marigold) [reply python-sgd-regression/sgd_regression.py]: PrettyPFA is neat, building regression model was indeed trivial.

The only problem with PrettyPFA is that titus does not support python 3... We have 3 options here (in order of my preference):
1. Translate PrettyPFA to PFA in Woken, i.e. `io_helper.save_results(pfa, '', 'pfa_pretty')`
2. Use (unofficial) python 3 fork https://github.com/animator/hadrian and convert it in the container
3. Use python 2 and convert it in the container

File: python-sgd-regression/sgd_regression.py (hunk 10 above)
Comment (ludovicc) [reply python-sgd-regression/sgd_regression.py]: Good point, we need to add PREVIOUS_JOB_ID to recover the previous model if any.

A verb is also sent to the container - compute (see https://github.com/LREN-CHUV/java-base-docker-images/blob/master/java-mip/Dockerfile#L50), but we can add variants, for example 'compute intermediate' and 'compute last'

File: python-sgd-regression/sgd_regression.py [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+
+    # Get existing results with partial model if they exist
+    job_result = io_helper.get_results()
+
+    if job_result:
+        # reconstruct estimator from metadata in PFA
+        pfa = json.loads(job_result.data)
+        estimator = deserialize_sklearn_estimator(pfa['metadata'])
+    else:
+        # TODO: SGD-type algorithms require normalized data! how do we do that in incremental learning?
```
Comment (Marigold) [inline python-sgd-regression/sgd_regression.py]: Normalization in incremental learning might be tricky, I'll take a shot at it tomorrow.

File: python-sgd-regression/sgd_regression.py (hunk 10 above)
Comment (Marigold) [reply python-sgd-regression/sgd_regression.py]: > but we can add variants, for example 'compute intermediate' and 'compute last'

Cool, `cmopute *` together with `PREVIOUS_JOB_ID` would definitely solve the problem.

File: python-sgd-regression/sklearn_to_pfa/sklearn_to_pfa.py [hunk 14]
```diff
@@ -0,0 +1,59 @@
+# TODO: move to a separate repository or at least python-mip
```
Comment (Marigold) [inline python-sgd-regression/sklearn_to_pfa/sklearn_to_pfa.py]: This is very, very rough basis for library for converting scikit-learn models into PrettyPFA. Again, I'm not sure where to put it - should I put it into `python-mip` or maybe create a new repo under `HBPMedical` (in that case, could you please tell me how are you using atomist for creating python packages?). Thanks!
